File: be-1-nestjs/src/auth/util.verify-token.ts
// https://github.com/awslabs/aws-support-tools/blob/master/Cognito/decode-verify-jwt/decode-verify-jwt.ts

import { promisify } from 'util';
import * as Axios from 'axios';
import * as jsonwebtoken from 'jsonwebtoken';
import * as jwkToPem from 'jwk-to-pem';
import { config } from '@/config'


export interface VerifyAccessTokenResult {
  readonly sub: string;
  readonly userName: string;
  readonly clientId: string;
  readonly isValid: boolean;
  readonly error?: any;
}

interface TokenHeader {
  kid: string;
  alg: string;
}
interface PublicKey {
  alg: string;
  e: string;
  kid: string;
  kty: string;
  n: string;
  use: string;
}
interface PublicKeyMeta {
  instance: PublicKey;
  pem: string;
}

interface PublicKeys {
  keys: PublicKey[];
}

interface MapOfKidToPublicKey {
  [key: string]: PublicKeyMeta;
}

interface Claim {
  sub: string;
  iss: string;
  version: number;
  client_id: string;
  origin_jti: string;
  token_use: string;
  scope: string;
  auth_time: number;
  exp: number;
  iat: number;
  jti: string;
  username: string;
}

const cognitoIssuer = () => {
  const { poolId, clientRegion } = config().cognito
  return `https://cognito-idp.${clientRegion}.amazonaws.com/${poolId}`;
}


let cacheKeys: MapOfKidToPublicKey | undefined;
const getPublicKeys = async (): Promise<MapOfKidToPublicKey> => {
  if (!cacheKeys) {
    const url = `${cognitoIssuer()}/.well-known/jwks.json`;
    console.log('...COGNITO --> getPublicKeys')
    const publicKeys = await Axios.default.get<PublicKeys>(url);
    cacheKeys = publicKeys.data.keys.reduce((agg, current) => {
      const pem = jwkToPem(current);
      agg[current.kid] = { instance: current, pem };
      return agg;
    }, {} as MapOfKidToPublicKey);
    return cacheKeys;
  } else {
    return cacheKeys;
  }
};

const verifyPromised = promisify(jsonwebtoken.verify.bind(jsonwebtoken));


export const verifyAccessToken = async (token: string): Promise<VerifyAccessTokenResult> => {
  let result: VerifyAccessTokenResult
  try {
    console.log(`...checking ${token.substring(0, 8)}...`);
    const tokenSections = (token || '').split('.');
    if (tokenSections.length < 2) {
      throw new Error('requested token is invalid');
    }
    const headerJSON = Buffer.from(tokenSections[0], 'base64').toString('utf8');
    const header = JSON.parse(headerJSON) as TokenHeader;
    const keys = await getPublicKeys();
    const key = keys[header.kid];
    if (key === undefined) {
      throw new Error('claim made for unknown kid');
    }
    const claim = await verifyPromised(token, key.pem) as Claim;
    const currentSeconds = Math.floor((new Date()).valueOf() / 1000);
    if (currentSeconds > claim.exp || currentSeconds < claim.auth_time) {
      throw new Error('claim is expired or invalid');
    }
    if (claim.iss !== cognitoIssuer()) {
      throw new Error('claim issuer is invalid');
    }
    if (claim.token_use !== 'access') {
      throw new Error('claim use is not access');
    }
    console.log(`...confirmed - ${claim.username} - ${claim.sub}`);
    result = { userName: claim.username, sub: claim.sub, clientId: claim.client_id, isValid: true };
  } catch (error) {
    result = { userName: '', sub: '', clientId: '', error, isValid: false };
  }
  return result;
};

File: be-2-flask/app/auth/util_verify_token.py
import base64
import json
import time
from typing import TypedDict
from jose import jwk, jwt
from jose.utils import base64url_decode

from .util_get_jwk import cognito_issuer, cognito_keys, JsonWebKey, JWKS
from .types import UserIdentity


class VerifyAccessTokenResult(TypedDict):
    sub: str
    user_name: str
    client_id: str
    is_valid: bool
    error: str


class TokenHeader(TypedDict):
    kid: str
    alg: str


class Claim(TypedDict):
    sub: str
    iss: str
    version: str
    client_id: str
    origin_jti: str
    token_use: str
    scope: str
    auth_time: str
    exp: str
    iat: str
    jti: str
    username: str


def create_identity(verify_result: VerifyAccessTokenResult) -> UserIdentity:
    return {
        "sub": verify_result["sub"],
        "userName": verify_result["user_name"],
        "clientId": verify_result["client_id"],
    }


def verify_access_token(token) -> VerifyAccessTokenResult:
    try:
        print(f"...checking {token[:8]}...")
        token_sections = (token or "").split(".")
        if len(token_sections) < 2:
            raise Exception("requested token is invalid")

        header_json = base64.urlsafe_b64decode(token_sections[0] + "===").decode("utf-8")
        header: TokenHeader = json.loads(header_json)
        kid: str = header["kid"]

        keys: JWKS = cognito_keys()
        key_index = -1
        for i in range(len(keys)):
            if kid == keys[i]["kid"]:
                key_index = i
                break
        key: JsonWebKey = keys[key_index]
        public_key = jwk.construct(key)
        message, encoded_signature = str(token).rsplit(".", 1)
        # decode the signature
        decoded_signature = base64url_decode(encoded_signature.encode("utf-8"))
        # verify the signature
        if not public_key.verify(message.encode("utf8"), decoded_signature):
            raise Exception("signature verification failed")

        claim: Claim = jwt.get_unverified_claims(token)

        current_seconds = int(time.time())
        if current_seconds > claim["exp"] or current_seconds < claim["auth_time"]:
            raise Exception("claim is expired or invalid")
        if claim["iss"] != cognito_issuer():
            raise Exception("claim issuer is invalid")
        if claim["token_use"] != "access":
            raise Exception("claim use is not access")

        exp_sec = int(claim["exp"] - time.time())
        print(f"...confirmed - {claim['username']} - {claim['sub']} - {exp_sec}sec")

        return {"user_name": claim["username"], "sub": claim["sub"], "client_id": claim["client_id"], "is_valid": True}
    except Exception as error:
        return {"user_name": "", "sub": "", "client_id": "", "error": str(error), "is_valid": False}
